torch/csrc/distributed/c10d/symm_mem/nvshmem_team_manager.hpp

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+#include <c10/cuda/CUDAException.h>
+#include <c10/util/Exception.h>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+// Starting from NVSHMEM 3.3.9, nvshmem_host.h exists so that we can cleanly
+// include only the nvshmem host library headers:
+// #include <nvshmem_host.h>
+// It translates into the following two lines:
+#include <host/nvshmem_api.h>
+#include <host/nvshmemx_api.h>
+// For maximum compatibility, we use the "host/" style for now.
+
+namespace c10d::nvshmem_extension {
+
+// This corresponds to max nblocks
+constexpr int MAX_N_TEAMS = 32;
+
+// A pool of teams for each group. These are duplicate teams.
+using nvshmemTeamPool_t = std::vector<nvshmem_team_t>;
+
+// Manage all the team business. Singleton.
+class TeamManager {
+ public:
+  // Get single, global manager.
+  static TeamManager& get() {
+    static TeamManager manager;
+    return manager;
+  }
+
+  // Get a team for a group.
+  nvshmem_team_t get_team(
+      const std::string& group_name,
+      const std::vector<int>& global_ranks) {
+    bool pool_updated = false; // unused
+    auto& team_pool =
+        group_to_team_pool(group_name, global_ranks, 1, pool_updated);
+    // Return the fist available team
+    return team_pool[0];
+  }
+
+  // Get n teams for a group.
+  // The first element of the returned pair is the team pool on host side.
+  // The second element of the returned pair is the team pool on device side.
+  std::pair<const nvshmemTeamPool_t&, nvshmem_team_t*> get_n_teams(
+      const std::string& group_name,
+      const std::vector<int>& global_ranks,
+      const int need_n) {
+    // Get the team pool with the requested number of teams
+    bool pool_updated = false;
+    auto& team_pool =
+        group_to_team_pool(group_name, global_ranks, need_n, pool_updated);
+    // Check if the pool already exists in device memory
+    nvshmem_team_t* team_pool_dev = nullptr;
+    constexpr auto pool_bytes = sizeof(nvshmem_team_t) * MAX_N_TEAMS;
+    auto it = team_pool_devptrs_.find(group_name);
+    if (it == team_pool_devptrs_.end()) {
+      // If not, allocate a new pool in device memory
+      C10_CUDA_CHECK(cudaMalloc((void**)&team_pool_dev, pool_bytes));
+      team_pool_devptrs_[group_name] = team_pool_dev;
```

Comment:
you should think about structuring the code in such a way that when group is destroyed, it's team manager entries are also freed, right now you have cudaMalloc (and possibly other resources) that's leaking.

Reply:
Normally yes (thanks for the suggestion)! But here I have specific comments + warning message, so I kind of prefer writing it out explicitly.

Reply:
@kwen2501 True, but comments has way more footguns than RAII